opaque_keys/edx/locator.py

```diff
@@ -119,8 +119,25 @@ def parse_url(cls, string):
         with key 'id' and optional keys 'branch' and 'version_guid'.
 
         Raises:
-            InvalidKeyError: if string cannot be parsed.
-        """
+            InvalidKeyError: if string cannot be parsed -or- string ends with a newline.
+        """
+        # URL_RE above cannot detect a single trailing newline, so guard against them
+        # by detecting them separately. An example to demonstrate:
+        #
+        # >>> import re
+        # >>> x = re.compile('(.+)$')
+        # >>> print x.match('foo')
```

Comment:
I mentioned elsewhere that PCRE has \z for matching end of string, not including newline.  Python seems to call that \Z.  Also, . doesn't match \n unless you enable /s (singleline)..  So can you do this as:

```
>>> import re
>>> x = re.compile('(.+)\Z')
>>> print x.match('foo')
<_sre.SRE_Match object at 0x10deeebe8>
>>> print x.match('foo\n')
None
```